File: godart.go
package main

import (
	"C"
	"fmt"
	"github.com/mraleph/go_dart_ffi_example/dart_api_dl"
	"time"
	"unsafe"
)

//export InitializeDartApi
func InitializeDartApi(api unsafe.Pointer) {
	dart_api_dl.Init(api)
}

//export StartWork
func StartWork(port int64, goFuncName *C.char) {
    funcName := C.GoString(goFuncName)
	fmt.Println("Go: Starting some asynchronous work")
	switch funcName {
        case "goInit":
            go goInit(port)
        case "goNetEnv":
            go goNetEnv(port)
        default:
            fmt.Printf("GO: goFuncName[%s] does not exist",funcName)
            dart_api_dl.SendToPort(port,  "default")
    }

	fmt.Println("Go: Returning to Dart")
}

func goInit(port int64) {
	var counter int64 = 50
    time.Sleep(2 * time.Second)
    fmt.Printf("GO: 2 seconds passed, goInit back: %d\n",counter)
    dart_api_dl.SendToPort(port,  "goInit")
}

func goNetEnv(port int64) {
    var counter int64 = 60
    time.Sleep(2 * time.Second)
    fmt.Printf("GO: 2 seconds passed, goNetEnv back: %d\n",counter)
    dart_api_dl.SendToPort(port,  "goNetEnv")
}



// func work(port int64) {
// 	var counter int64
// 	for {
// 		time.Sleep(2 * time.Second)
// 		fmt.Println("GO: 2 seconds passed")
// 		counter++
// 		dart_api_dl.SendToPort(port, counter)
// 	}
// }

// Unused
func main() {}

File: godart.dart
import 'dart:ffi';
import 'dart:isolate';
import 'package:ffi/ffi.dart';

typedef StartWorkType = Void Function(Int64 port, Pointer<Utf8> goFuncName);
typedef StartWorkFunc = void Function(int port, Pointer<Utf8> goFuncName);

class Work extends Struct {
  Pointer<Utf8> name;
}

class NativeGoFunc {
  static const String goInit = 'goInit';
  static const String goNetEnv = 'goNetEnv';
}

void main() async {
  print('-----------111--------');
  testGo().then((value) {
    print('------$value---------');
    final utf = Pointer<Utf8>.fromAddress(value as int);
    print('------${utf.toDartString()}---------');

    // final work = Pointer<Work>.fromAddress(value as int);
    // print(work.ref.name.toDartString());
  });
  print('-----------222--------');
}

Future<dynamic> testGo() {
  final lib = DynamicLibrary.open("./godart.so");

  final initializeApi =
      lib.lookupFunction<IntPtr Function(Pointer<Void>), int Function(Pointer<Void>)>(
          "InitializeDartApi");
  if (initializeApi(NativeApi.initializeApiDLData) != 0) {
    throw "Failed to initialize Dart API";
  }

  // final interactiveCppRequests = ReceivePort()
  //   ..listen((data) {
  //     print('Received: ${data} from Go');
  //     receivePort.close();
  //     return data;
  //   });
  // final int nativePort = interactiveCppRequests.sendPort.nativePort;

  final ReceivePort receivePort = ReceivePort();
  final int nativePort = receivePort.sendPort.nativePort;

  print("nativePort: ${nativePort}");

  final StartWorkFunc startWork =
      lib.lookup<NativeFunction<StartWorkType>>("StartWork").asFunction();

  startWork(nativePort, NativeGoFunc.goInit.toNativeUtf8());
  return receivePort.first;
}
